state/modelgeneration.go

```diff
@@ -424,6 +422,52 @@ func (g *Generation) Commit(userName string) (int, error) {
 	return newGenId, nil
 }
 
+// assignedWithAllUnits generates a new value for the branch's
+// AssignedUnits field, to indicate the all units of changed applications
+// are tracking the branch.
+func (g *Generation) assignedWithAllUnits() (map[string][]string, error) {
+	assigned := g.AssignedUnits()
+	for app := range assigned {
+		units, err := appUnitNames(g.st, app)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		assigned[app] = units
+	}
+	return assigned, nil
+}
+
+// commitConfigTxnOps iterates over all the applications with configuration
+// deltas, determines their effective new settings, then gathers the
+// operations representing the changes so that they can all be applied in a
+// single transaction.
+func (g *Generation) commitConfigTxnOps() ([]txn.Op, error) {
+	var ops []txn.Op
+	for appName, delta := range g.Config() {
+		if len(delta) == 0 {
+			continue
+		}
+		app, err := g.st.Application(appName)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		// Apply the branch delta to the application's charm config settings.
+		cfg, err := readSettings(g.st.db(), settingsC, app.charmConfigKey())
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		cfg.applyChanges(delta)
+
+		// Assert that the settings document has not changed underneath us
+		// in addition to appending the field changes.
+		ops = append(ops, cfg.assertUnchangedOp())
```

Comment:
Isn't it possible that the underlying settings have been updated to match the operations for a particular application?
In those situations, isn't it possible that there is a no-op for updating the settings for that application?

Reply:
Yes, true. I will omit the check when there are no settings modification ops generated.